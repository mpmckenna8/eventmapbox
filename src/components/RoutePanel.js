import PropTypes from 'prop-types';
import React, {Component} from 'react';
import {connect} from 'react-redux';
import RouteElevation from './RouteElevation';

class RoutePanel extends Component {
  render() {
    return (
      <div className='relative mt6 bg-white w-full w420-mm hmin42 hmin48-mm shadow-darken25 flex-parent flex-parent--column flex-parent--center-main'>
        {this.routeInfo()}
        {
          this.props.modality === 'bike'
          ? <RouteElevation
              route={this.props.route}
              accessToken={this.props.mapboxAccessToken}
            />
          : null
        }
      </div>
    );
  }

  routeInfo() {
    if (this.props.route) {
      return (
        <div className='flex-parent flex-parent--row flex-parent--center-cross'>
          <div className='w48 hmin48 flex-parent flex-parent--center-main flex-parent--center-cross'>
            <svg className='icon color-gray'><use xlinkHref={'#icon-' + this.props.modality}></use></svg>
          </div>
          <div className='bottom pr12'>
            {this.secondsToReadableTime(this.props.route.duration)}
            {this.props.modality === 'car' ? ' with current traffic' : ''}
          </div>
          <div className='bottom color-gray'>
            {(this.props.route.distance / 1000).toFixed(1)} km
          </div>
        </div>
      );
    } else if (this.props.routeStatus === 'pending') {
      return <div className='loading loading--s'/>;
    } else if (this.props.routeStatus === 'error') {
      return <div className='txt-s txt-m-mm animation-subtle-shake px18'>Sorry, no route found for these locations.</div>;
    } else return null;
  }



  secondsToReadableTime(duration) {
    let minutes = Math.max(1, Math.round(duration / 60));

    let hours = Math.floor(minutes / 60);
    minutes -= 60 * hours;

    let days = Math.floor(hours / 24);
    hours -= 24 * days;

    const daysString = (days > 0) ? `${days} day${days > 1 ? 's,' : ','} ` : '';
    const hoursString = (hours > 0) ? `${hours} hour${hours > 1 ? 's and' : ' and'} ` : '';
    const minutesString = `${minutes} minute${minutes > 1 ? 's' : ''} `;
    return daysString + hoursString + minutesString;
  }
}

RoutePanel.propTypes = {
  mapboxAccessToken: PropTypes.string,
  modality: PropTypes.string,
  route: PropTypes.object,
  routeStatus: PropTypes.string.isRequired,
};

const mapStateToProps = (state) => {
  return {
    mapboxAccessToken: state.app.mapboxAccessToken,
    modality: state.app.modality,
    route: state.app.route,
    routeStatus: state.app.routeStatus,
  };
};

export {RoutePanel};
export default connect(mapStateToProps)(RoutePanel);
