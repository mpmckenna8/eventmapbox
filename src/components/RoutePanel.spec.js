import React from 'react';
import renderer from 'react-test-renderer';
import {RoutePanel} from './RoutePanel';
import {bikeRoute, walkingRoute} from './fixtures/RoutePanel.fixtures.js';

it('bike route renders correctly', () => {

  const component = renderer.create(
    <RoutePanel
      mapboxAccessToken='test'
      modality='bike'
      route={bikeRoute}
      routeStatus='idle'
    />
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('walking route renders correctly', () => {

  const component = renderer.create(
    <RoutePanel
      mapboxAccessToken='test'
      modality='walk'
      route={walkingRoute}
      routeStatus='idle'
    />
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('pending route renders correctly', () => {

  const component = renderer.create(
    <RoutePanel
      mapboxAccessToken='test'
      modality='walk'
      route={null}
      routeStatus='pending'
    />
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('error route renders correctly', () => {

  const component = renderer.create(
    <RoutePanel
      mapboxAccessToken='test'
      modality='walk'
      route={null}
      routeStatus='error'
    />
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
