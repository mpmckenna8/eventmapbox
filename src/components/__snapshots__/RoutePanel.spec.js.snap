exports[`test bike route renders correctly 1`] = `
<div
  className="relative mt6 bg-white w-full w420-mm hmin42 hmin48-mm shadow-darken25 flex-parent flex-parent--column flex-parent--center-main">
  <div
    className="flex-parent flex-parent--row flex-parent--center-cross">
    <div
      className="w48 hmin48 flex-parent flex-parent--center-main flex-parent--center-cross">
      <svg
        className="icon color-gray">
        <use
          xlinkHref="#icon-bike" />
      </svg>
    </div>
    <div
      className="bottom pr12">
      17 minutes 
      
    </div>
    <div
      className="bottom color-gray">
      2.6
       km
    </div>
  </div>
  <div
    className="loading--s" />
</div>
`;

exports[`test error route renders correctly 1`] = `
<div
  className="relative mt6 bg-white w-full w420-mm hmin42 hmin48-mm shadow-darken25 flex-parent flex-parent--column flex-parent--center-main">
  <div
    className="txt-s txt-m-mm animation-subtle-shake px18">
    Sorry, no route found for these locations.
  </div>
</div>
`;

exports[`test pending route renders correctly 1`] = `
<div
  className="relative mt6 bg-white w-full w420-mm hmin42 hmin48-mm shadow-darken25 flex-parent flex-parent--column flex-parent--center-main">
  <div
    className="loading loading--s" />
</div>
`;

exports[`test walking route renders correctly 1`] = `
<div
  className="relative mt6 bg-white w-full w420-mm hmin42 hmin48-mm shadow-darken25 flex-parent flex-parent--column flex-parent--center-main">
  <div
    className="flex-parent flex-parent--row flex-parent--center-cross">
    <div
      className="w48 hmin48 flex-parent flex-parent--center-main flex-parent--center-cross">
      <svg
        className="icon color-gray">
        <use
          xlinkHref="#icon-walk" />
      </svg>
    </div>
    <div
      className="bottom pr12">
      8 days, 4 hours and 41 minutes 
      
    </div>
    <div
      className="bottom color-gray">
      982.4
       km
    </div>
  </div>
</div>
`;
